### PR TITLE
Add SBOM manifest to nuget packages

### PR DESCRIPTION
**Pull request by PaulHigin**
This PR adds SBOM manifests for created nuget packages for coordinated package builds.

SBOM manifests are added by splitting the existing `New-ILNugetPacakge` function, which creates all nuget packages at once, into `New-ILNugetPackageSource` and `New-ILNugetPackageFromSource` functions.  These new functions create source and package for just one file at a time, so that an SBOM manifest can be created (via yml template) for each file/package.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: tools/packaging/packaging.psm1 [hunk 1]
```diff
@@ -1710,14 +1713,321 @@ function CreateNugetPlatformFolder
...
+            {
+                $deps.Add([tuple]::Create([tuple]::Create('id', $packageInfo.Name), [tuple]::Create('version', $packageInfo.Version))) > $null
+            }
+        }
+    }
+
+    $srcFilePackagePath = Join-Path $PackagePath $fileBaseName
+
+    Write-Verbose -Verbose "New-ILNugetPackageFromSource: Creating nuget package for file: $FileName from source path: $srcFilePackagePath"
+    Write-Verbose -Verbose -Message "$(Get-ChildItem -Path $srcFilePackagePath)"
```
Comment (JamesWTruher) [inline tools/packaging/packaging.psm1:2014]: Are usually prefer something more like this:
```powershell
PS> (ls).foreach({write-verbose -verbose -message $_})
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/out
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/PSUnixUtilCompleters
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/tests
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/build.ps1
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/completions.gif
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/LICENSE
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/OnStart.ps1
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/PSUnixUtilCompleters.psd1
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/README.md
```
because this is a little messy
```powershell
PS> write-verbose -verbose "$(ls)"                          
VERBOSE: /Users/james/src/github/forks/jameswtruher/UnixCompleters/out /Users/james/src/github/forks/jameswtruher/UnixCompleters/PSUnixUtilCompleters /Users/james/src/github/forks/jameswtruher/UnixCompleters/tests /Users/james/src/github/forks/jameswtruher/UnixCompleters/build.ps1 /Users/james/src/github/forks/jameswtruher/UnixCompleters/completions.gif /Users/james/src/github/forks/jameswtruher/UnixCompleters/LICENSE /Users/james/src/github/forks/jameswtruher/UnixCompleters/OnStart.ps1 /Users/james/src/github/forks/jameswtruher/UnixCompleters/PSUnixUtilCompleters.psd1 /Users/james/src/github/forks/jameswtruher/UnixCompleters/README.md
```

File: tools/packaging/packaging.psm1 [hunk 2]
```diff
@@ -1710,14 +1713,321 @@ function CreateNugetPlatformFolder
...
+
+    if (! (Test-Path $srcFilePackagePath)) {
+        $msg = "Expected nuget source path $srcFilePackagePath for file $fileBaseName does not exist."
+        Write-Verbose -Verbose $msg
```
Comment (JamesWTruher) [inline tools/packaging/packaging.psm1:2018]:
```suggestion
        Write-Verbose -Verbose -Message $msg
```

File: tools/releaseBuild/azureDevOps/templates/nuget-pkg-sbom.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- ${{ each value in parameters.ListOfFiles }}:
+  - pwsh: |
+      $FileName = '${{ value }}'
+      $FileBaseName = [System.IO.Path]::GetFileNameWithoutExtension($FileName)
+      $FilePackagePath = Join-Path -Path '${{ parameters.PackagePath }}' -ChildPath $FileBaseName
+      Write-Verbose -Verbose "FileName to package: $FileName"
+      Write-Verbose -Verbose "FilePackage path: $FilePackagePath"
+      # Set SBOM package name
+      $vstsCommandString = "vso[task.setvariable variable=SbomFilePackageName]${FileBaseName}"
+      Write-Host "sending " + $vstsCommandString
```
Comment (JamesWTruher) [inline tools/releaseBuild/azureDevOps/templates/nuget-pkg-sbom.yml:32]:
```suggestion
      Write-Verbose -Verbose -Message "sending $vstsCommandString"
```
???

**Review by JamesWTruher — APPROVED** [review]
just a couple of suggestions but otherwise looks good

File: tools/packaging/packaging.psm1 (hunk 1 above)
Comment (PaulHigin) [reply tools/packaging/packaging.psm1:2014]: This is actually not needed any more and I'll remove it.

File: tools/releaseBuild/azureDevOps/templates/nuget.yml [hunk 4]
```diff
@@ -10,7 +10,7 @@ jobs:
   pool:
     name: PowerShell1ES
     demands:
-    - ImageOverride -equals MMS2019TLS
+    - ImageOverride -equals MMS2019
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/nuget.yml:13]:
```suggestion
    - ImageOverride -equals MMS2019TLS
```

We shouldn't revert this change

File: tools/releaseBuild/azureDevOps/templates/nuget-pkg-sbom.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      New-ILNugetPackageSource -File $FileName -PackagePath '${{ parameters.PackagePath }}' -PackageVersion '${{ parameters.PackageVersion }}' -WinFxdBinPath '${{ parameters.WinFxdPath }}' -LinuxFxdBinPath '${{ parameters.LinuxFxdPath }}' -GenAPIToolPath '${{ parameters.GenAPIToolPath }}'
+    displayName: 'Create NuGet Package source for single file'
+
+  - template: Sbom.yml@ComplianceRepo
+    parameters:
+      BuildDropPath: $(SbomFilePackagePath)
+      Build_Repository_Uri: 'https://github.com/powershell/powershell'
+      PackageName: $(SbomFilePackageName)
+      PackageVersion: ${{ parameters.PackageVersion }}
+      sourceScanPath: $(SbomFilePackagePath)
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/nuget-pkg-sbom.yml:51]: CGManifest shouldn't be in the drop, not sure this is bracksing

File: tools/packaging/packaging.psm1 [hunk 6]
```diff
@@ -1710,14 +1713,393 @@ function CreateNugetPlatformFolder
...
+            $deps.Add([tuple]::Create([tuple]::Create('id', 'Microsoft.PowerShell.ConsoleHost'), [tuple]::Create('version', $PackageVersion))) > $null
+            $deps.Add([tuple]::Create([tuple]::Create('id', 'Microsoft.PowerShell.Security'), [tuple]::Create('version', $PackageVersion))) > $null
+            $deps.Add([tuple]::Create([tuple]::Create('id', 'System.Management.Automation'), [tuple]::Create('version', $PackageVersion))) > $null
+            foreach($packageInfo in (Get-ProjectPackageInformation -ProjectName $FileBaseName))
+            {
+                $deps.Add([tuple]::Create([tuple]::Create('id', $packageInfo.Name), [tuple]::Create('version', $packageInfo.Version))) > $null
+            }
+            $deps.Add([tuple]::Create([tuple]::Create('id', 'Microsoft.WSMan.Management'), [tuple]::Create('version', $PackageVersion))) > $null
+            $deps.Add([tuple]::Create([tuple]::Create('id', 'Microsoft.PowerShell.Commands.Diagnostics'), [tuple]::Create('version', $PackageVersion))) > $null
+            $deps.Add([tuple]::Create([tuple]::Create('id', 'Microsoft.Management.Infrastructure.CimCmdlets'), [tuple]::Create('version', $PackageVersion))) > $null
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1:1838]: Is it possible to have one data structure of dependecies?

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
Please give your thoughts on the comments I left

File: tools/releaseBuild/azureDevOps/templates/nuget-pkg-sbom.yml (hunk 5 above)
Comment (PaulHigin) [reply tools/releaseBuild/azureDevOps/templates/nuget-pkg-sbom.yml:51]: The file is removed before the package is created, but is there when the SBOM is generated.  Not sure if that is a problem, but I can put it in a different location.

File: tools/packaging/packaging.psm1 (hunk 6 above)
Comment (PaulHigin) [reply tools/packaging/packaging.psm1:1838]: This is easy and quick to generate so I feel it is not a problem wrapping it in a helper function.

File: tools/packaging/packaging.psm1 [hunk 7]
```diff
@@ -2294,6 +2685,8 @@ function CleanupGeneratedSourceCode
         '[Microsoft.PowerShell.Commands.SetStrictModeCommand.ArgumentToPSVersionTransformationAttribute]'
         '[Microsoft.PowerShell.Commands.HttpVersionCompletionsAttribute]'
         '[System.Management.Automation.ArgumentToVersionTransformationAttribute]'
+        '[Microsoft.PowerShell.Commands.InvokeCommandCommand.ArgumentToPSVersionTransformationAttribute]'
+        '[Microsoft.PowerShell.Commands.InvokeCommandCommand.ValidateVersionAttribute]'
         )
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1:2690]: This PR needs to be backported, please move these changes to another PR

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: tools/packaging/packaging.psm1 (hunk 7 above)
Comment (PaulHigin) [reply tools/packaging/packaging.psm1:2690]: @TravisEz13   It is only two lines, and I need the changes in order to test.  I'll create a separate PR for these two lines, but in the meantime I think this PR can be approved.

Comment (PaulHigin) [reply tools/packaging/packaging.psm1:2690]: https://github.com/PowerShell/PowerShell/pull/16800

**Review by TravisEz13 — APPROVED** [review]
We should have @adityapatwardhan  sign off too

**Comment by PaulHigin** [comment]
@adityapatwardhan   Ping...

File: tools/packaging/packaging.psm1 [hunk 8]
```diff
@@ -1710,14 +1713,397 @@ function CreateNugetPlatformFolder
...
+["Id", "Microsoft.PowerShell.SecretStore"], ["Version", "1.1.1.0"]
+#>
+function New-CGManifest
+{
+    param (
+        [parameter(Mandatory = $true)]
+        [string] $FilePath,
+
+        [parameter(Mandatory = $false)]
+        [tuple[ [tuple[string, string]], [tuple[string, string]] ] []] $Dependencies
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1:1745]: nit: Maybe a comment about the tuple structure

File: tools/packaging/packaging.psm1 [hunk 9]
```diff
@@ -1792,185 +2183,182 @@ function New-ILNugetPackage
...
             }
 
-            #region nuspec
-            # filed a tracking bug for automating generation of dependecy list: https://github.com/PowerShell/PowerShell/issues/6247
-            $deps = [System.Collections.ArrayList]::new()
+            Write-Log "Copied the built-in modules to contentFiles for the SDK package"
+        }
+
+        #region nuspec
+        # filed a tracking bug for automating generation of dependecy list: https://github.com/PowerShell/PowerShell/issues/6247
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1:2261]: Is this code same as L1738? Can we refactor to reuse common code?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: tools/packaging/packaging.psm1 (hunk 9 above)
Comment (PaulHigin) [reply tools/packaging/packaging.psm1:2261]: The `New-ILNugetPackage` function is no longer used, however I didn't want to remove it during testing.  But I'll go ahead and remove it now.

File: tools/packaging/packaging.psm1 (hunk 8 above)
Comment (PaulHigin) [reply tools/packaging/packaging.psm1:1745]: This is commented above at line 1734.

**Review by adityapatwardhan — APPROVED** [review]
